### PR TITLE
Implement PIP messages, request builder, and handlers

### PR DESCRIPTION
**Pull request by rphmeier**
`OnDemand` (and by extension RPC) still uses these messages inefficiently. A more sophisticated dispatch mechanism will come in a follow-up PR.

Should be merged after #4934 

Information for reviewers:
  - Previously, `ethcore-light` implemented a bastardized version of LES.
  - This is the first step towards migrating to [PIP](https://github.com/paritytech/parity/wiki/The-Parity-Light-Protocol-(PIP)) by moving to heterogeneous back-referencing requests
  - The bulk of the code changed here is just request definitions coupled with encoding boilerplate.

### CODE REVIEW

File: ethcore/light/src/net/request_credits.rs [hunk 1]
```diff
@@ -214,61 +215,34 @@ impl FlowParams {
...
-
-		let start = credits.current();
-
-		if start <= cost.0 {
-			return 0;
-		} else if cost.1 == U256::zero() {
-			return usize::MAX;
+	pub fn compute_cost(&self, request: &Request) -> U256 {
+		match *request {
+			Request::Headers(ref req) => self.costs.headers * req.max.into(),
```
Comment (tomusdrw) [inline ethcore/light/src/net/request_credits.rs]: Probably would be better to use saturating_ops here?

File: ethcore/light/src/net/request_credits.rs [hunk 2]
```diff
@@ -214,61 +215,34 @@ impl FlowParams {
...
-		let max = (start - cost.0) / cost.1;
-		if max >= usize::MAX.into() {
-			usize::MAX
-		} else {
-			max.as_u64() as usize
-		}
+	/// Compute the cost of a set of requests.
+	/// This is the base cost plus the cost of each individual request.
+	pub fn compute_cost_multi(&self, requests: &[Request]) -> U256 {
+		requests.iter().fold(self.costs.base, |cost, req| cost + self.compute_cost(req))
```
Comment (tomusdrw) [inline ethcore/light/src/net/request_credits.rs]: `saturating_add`?

**Comment by debris** [comment]
merged with master to trigger the CI build

**Comment by arkpar** [comment]
Looks good overall. To huge too look into details.

**Pull request merged**